sdb/commands/linux/slabs.py

```diff
@@ -0,0 +1,141 @@
+#
+# Copyright 2019 Delphix
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# pylint: disable=missing-docstring
+
+import argparse
+from typing import Any, Callable, Dict, Iterable, List
+
+import drgn
+
+import sdb
+from sdb.commands.internal.fmt import size_nicenum
+from sdb.commands.internal.table import Table
+from sdb.commands.linux.internal import slub_helpers as slub
+
+
+class Slabs(sdb.Locator, sdb.PrettyPrinter):
+    # pylint: disable=too-few-public-methods
+
+    names = ["slabs"]
+
+    input_type = "struct kmem_cache *"
+    output_type = "struct kmem_cache *"
+
+    def _init_argparse(self, parser: argparse.ArgumentParser) -> None:
+        parser.add_argument(
+            '-H',
+            action='store_false',
+            help=
+            'do not display headers and separate fields by a single tab (scripted mode)'
+        )
+        parser.add_argument(
+            '-o',
+            metavar="PROPERTIES",
+            help='comma-separated list of properties to display')
+        parser.add_argument('-p',
+                            action='store_true',
+                            help='display numbers in parseable (exact) values')
+        parser.add_argument(
+            '-r',
+            '--recursive',
+            action='store_true',
+            help='recurse down children caches and print statistics')
+        parser.add_argument('-s',
+                            metavar="PROPERTY",
+                            help='sort rows by PROPERTY')
```

Comment:
I'm not sure what's expected in this case, but I was experimenting with this command and noticed the following inconsistency:
```
> slabs -s mem_used ! head -n3
name                        entry_size   objs mem_used util
--------------------------- ---------- ------ -------- ----
kmalloc-512                        512 239296  116.8MB   51

> slabs -s mem_used | head 1 | pp
name             entry_size objs mem_used util
---------------- ---------- ---- -------- ----
t10_pr_reg_cache        704    0     0.0B    0
```

It seems as though sorting only works when pretty printing, and not what the objects are piped into another SDB command. Is this a bug, or working as designed?

Reply:
Hey Prakash! This is a very interesting point that you brought up and it definitely has to do with how things are designed (specifically the Locator and PrettyPrinter hybrids bundled together with the nature of this command).

Since the command is a `Locator`+`PrettyPrinter` hybrid, it's functionality is currently structured in 2 parts:
[1] The `Locator` part which iterates through the slab caches in the order that we find them in the kernel
[2] The `PrettyPrinter` part which is drains the input that it gets from [1] (or whatever it is passed through the pipeline) in the `for obj in objs:` loop in the beginning of its method, and once that is done it does the sorting by itself and outputs everything.

With this in mind what you have on your example below makes sense:
```
> slabs -s mem_used | head 1 | pp
name             entry_size objs mem_used util
---------------- ---------- ---- -------- ----
t10_pr_reg_cache        704    0     0.0B    0
```
`sdb` figures out that the `slabs` command is not the last one in the pipeline and executes only the `Locator` functionality (part [1] above). The `-s` option is basically discarded as we never get to the pretty-printer part.

I wouldn't say that's a bug but more of a design trade-off. By not sorting in the `Locator` itself we "hold" one `drgn.Object` in memory (a slab cache) that we `yield` along the pipeline to each invocation making each yield `O(1)` in runtime. If we were to sort in the `Locator` and then `yield` each object we'll be holding all the `drgn.Object`s for all slab caches in memory until the `Locator` is fully drained. This would also make the runtime of the first invocation `O(nlogn)` but subsequent ones would still be `O(1)`.

If that trade-off is worth it for `slabs` I'm not sure. I do agree that `-s` being discarded may be misleading. At the same time I'm thinking that there could definitely be other commands in the future where memory could be a problem (e.g. a `pages` `Locator` that walks through all the physical memory pages in the system). And then the trade-off of consistent user experience comes in - do we make things like `-s` discarded for all commands to be consistent or we let each command do its own thing? Maybe the latter is not bad. At the same time I can see us keeping the discarding of options like `-s` consistent and maybe introducing a `sort` command where we leave it to the user to sort the objects in the pipeline in a more explicit way ( this would be similar to how we implemented `filter` instead of adding an option to each command that we want to filter the output from).

I'll  created an issue about this where we can discuss this separately. Definitely a great point.

Reply:
That all makes sense from an implementation perspective; and that's what I assumed was the culprit. I still think it's awkward from a UX perspective. With that said, I'm OK going with this as-is, and perhaps revisiting this in the future.

Creating some type of `sort` command, similar to `filter` seems reasonable to me, but then it raises the question of how do we do pretty printing of the sorted information.. e.g. would we still support `slabs -s`, or would we require `slabs | sort | pp` ?

We can continue this discussion in another issue, and/or at a later time. I'm OK moving forward with what you have here.

Reply:
IMO, if I request sorting by passing `-s ...`, the output should be sorted.  The alternative would be for the command to fail, with an error message like "I don't believe you really want to wait for the output to be sorted".  Yes, sorting uses CPU and memory, but if the user asks for it, we should do it.  Folks using a kernel debugger are typically sophisticated enough to figure out what's going on.  If the user gets tired of waiting they can hit ^c and then run it without the `-s`.

The default behavior (when `-s` is not specified) is debatable.  Perhaps it could be unsorted when used in the pipeline but sorted by `mem_used` when pretty-printing.

Reply:
I opened https://github.com/delphix/sdb/issues/40